### PR TITLE
Feature/glblagctwo 1092

### PR DESCRIPTION
**Pull request by mavdi**
<a href='#crh-start'></a><a href='#crh-data-%7B%22approved%22%3A%20%7B%22https%3A//github.com/ssebro%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%7D%7D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/agco/harvesterjs/pull/144%23issuecomment-145856690%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222015-10-06T13%3A31%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ssebro%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2098f02e0701f60843122af55a44df13aa168942f1%20lib/multi-sse.js%2021%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41173349%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22seems%20like%20a%20request%20validation%20middleware%2C%20which%20is%20nice%20%3A%29.%20Might%20be%20more%20obvious%20if%20you%20called%20it%20requestValidationMiddleware%20-%20do%20you%20agree%20that%20this%20change%20increases%20readability%3F%22%2C%20%22created_at%22%3A%20%222015-10-05T17%3A40%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ssebro%22%7D%7D%2C%20%7B%22body%22%3A%20%22sure%20yeah%20makes%20sense%2C%20I%27ll%20change%20it%22%2C%20%22created_at%22%3A%20%222015-10-05T18%3A07%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/767675%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mavdi%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/multi-sse.js%3AL1-129%22%7D%2C%20%22Pull%2098f02e0701f60843122af55a44df13aa168942f1%20lib/multi-sse.js%20109%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41177493%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22how%20are%20errors%20handled%20here%3F%22%2C%20%22created_at%22%3A%20%222015-10-05T18%3A13%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ssebro%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/multi-sse.js%3AL1-129%22%7D%2C%20%22Pull%2098f02e0701f60843122af55a44df13aa168942f1%20lib/multi-sse.js%20112%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41177637%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Definitely%20log%20something%20here%22%2C%20%22created_at%22%3A%20%222015-10-05T18%3A14%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ssebro%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/multi-sse.js%3AL1-129%22%7D%2C%20%22Pull%2098f02e0701f60843122af55a44df13aa168942f1%20lib/multi-sse.js%2079%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41177583%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Looks%20like%20you%27ll%20read%20continuously%20and%20rely%20on%20highland%20to%20allow%20pausing%20-%20is%20that%20correct%3F%22%2C%20%22created_at%22%3A%20%222015-10-05T18%3A14%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ssebro%22%7D%7D%2C%20%7B%22body%22%3A%20%22This%20line%20will%20feed%20the%20stream%20if%20the%20value%20we%20read%20isn%27t%20null%2C%20down%20the%20line%20we%20won%27t%20process%20the%20next%20event%20unless%20the%20one%20at%20hand%20is%20processed%20and%20sent%20through%2C%20i.e.%20when%20%60%60%60next%28%29%60%60%60%20is%20called%22%2C%20%22created_at%22%3A%20%222015-10-05T18%3A18%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/767675%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mavdi%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/multi-sse.js%3AL1-129%22%7D%2C%20%22Pull%2098f02e0701f60843122af55a44df13aa168942f1%20lib/multi-sse.js%2062%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41174158%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22not%20sure%20if%20me%20as%20this%20is%20a%20typical%20convention.%20I%27ve%20seen%20_this%20%26%20that%2C%20but%20not%20me.%20This%20is%20partially%20me%20wanting%20to%20keep%20things%20similar%2C%20and%20also%20me%20being%20curious%20about%20whether%20this%20is%20a%20new%20convention.%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-10-05T17%3A47%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/372075%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ssebro%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%20new%20convention.%20I%20normally%20do%20it%20as%20%60%60%60me%60%60%60%20but%20happy%20to%20switch%20to%20anything%20we%20normally%20use%20in%20codebase.%22%2C%20%22created_at%22%3A%20%222015-10-05T18%3A08%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/767675%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mavdi%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/multi-sse.js%3AL1-129%22%7D%7D%2C%20%22processed%22%3A%20%5B%22https%3A//github.com/agco/harvesterjs/pull/144%23issuecomment-145856690%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41173349%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41174158%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41176680%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41176800%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41177493%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41177583%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41177637%22%2C%20%22https%3A//github.com/agco/harvesterjs/pull/144%23discussion_r41178079%22%5D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/ssebro'></a>

- [ ] <a href='#crh-comment-Pull 98f02e0701f60843122af55a44df13aa168942f1 lib/multi-sse.js 21'></a> &nbsp;<b><a href='https://github.com/agco/harvesterjs/pull/144#discussion_r41173349'>File: lib/multi-sse.js:L1-129</a></b>
- <a href='https://github.com/ssebro'></a> seems like a request validation middleware, which is nice :). Might be more obvious if you called it requestValidationMiddleware - do you agree that this change increases readability?
- <a href='https://github.com/mavdi'></a> sure yeah makes sense, I'll change it
- [ ] <a href='#crh-comment-Pull 98f02e0701f60843122af55a44df13aa168942f1 lib/multi-sse.js 62'></a> &nbsp;<b><a href='https://github.com/agco/harvesterjs/pull/144#discussion_r41174158'>File: lib/multi-sse.js:L1-129</a></b>
- <a href='https://github.com/ssebro'></a> not sure if me as this is a typical convention. I've seen _this & that, but not me. This is partially me wanting to keep things similar, and also me being curious about whether this is a new convention.
- <a href='https://github.com/mavdi'></a> No new convention. I normally do it as ```me``` but happy to switch to anything we normally use in codebase.
- [ ] <a href='#crh-comment-Pull 98f02e0701f60843122af55a44df13aa168942f1 lib/multi-sse.js 109'></a> &nbsp;<b><a href='https://github.com/agco/harvesterjs/pull/144#discussion_r41177493'>File: lib/multi-sse.js:L1-129</a></b>
- <a href='https://github.com/ssebro'></a> how are errors handled here?
- [ ] <a href='#crh-comment-Pull 98f02e0701f60843122af55a44df13aa168942f1 lib/multi-sse.js 79'></a> &nbsp;<b><a href='https://github.com/agco/harvesterjs/pull/144#discussion_r41177583'>File: lib/multi-sse.js:L1-129</a></b>
- <a href='https://github.com/ssebro'></a> Looks like you'll read continuously and rely on highland to allow pausing - is that correct?
- <a href='https://github.com/mavdi'></a> This line will feed the stream if the value we read isn't null, down the line we won't process the next event unless the one at hand is processed and sent through, i.e. when ```next()``` is called
- [ ] <a href='#crh-comment-Pull 98f02e0701f60843122af55a44df13aa168942f1 lib/multi-sse.js 112'></a> &nbsp;<b><a href='https://github.com/agco/harvesterjs/pull/144#discussion_r41177637'>File: lib/multi-sse.js:L1-129</a></b>
- <a href='https://github.com/ssebro'></a> Definitely log something here

<a href='https://www.codereviewhub.com/agco/harvesterjs/pull/144?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/agco/harvesterjs/pull/144?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/agco/harvesterjs/pull/144?approve=0'></a>&nbsp;<a href='https://github.com/agco/harvesterjs/pull/144'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by ssebro** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/agco/harvesterjs/pull/144'>Refresh</a>.<a href='#crh-update-none'></a>

File: lib/multi-sse.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+MultiSSE.prototype.init = function(harvesterApp) {
+    this.options = harvesterApp.options;
+    this.db = mongojs(this.options.oplogConnectionString);
+    this.coll = this.db.collection('oplog.rs');
+    this.harvesterApp = harvesterApp;
+    harvesterApp.router.get(this.options.baseUrl + '/changes/stream', this.checkValidRequestMiddleware.bind(this), tinySSE.head(), tinySSE.ticker({seconds: 3}), this.handler.bind(this));
+};
+
+MultiSSE.prototype.checkValidRequestMiddleware = function (req, res, next) {
```
Comment (ssebro) [inline lib/multi-sse.js]: seems like a request validation middleware, which is nice :). Might be more obvious if you called it requestValidationMiddleware - do you agree that this change increases readability?

File: lib/multi-sse.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                detail: 'Could not parse the time stamp provided'
+            });
+        }
+    }
+
+    next();
+};
+
+MultiSSE.prototype.handler = function (req, res, next) {
+    var me = this;
```
Comment (ssebro) [inline lib/multi-sse.js]: not sure if me as this is a typical convention. I've seen _this & that, but not me. This is partially me wanting to keep things similar, and also me being curious about whether this is a new convention.

File: lib/multi-sse.js (hunk 1 above)
Comment (mavdi) [inline lib/multi-sse.js]: sure yeah makes sense, I'll change it

File: lib/multi-sse.js (hunk 2 above)
Comment (mavdi) [inline lib/multi-sse.js]: No new convention. I normally do it as ```me``` but happy to switch to anything we normally use in codebase.

File: lib/multi-sse.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                var model = me.harvesterApp.adapter.model(routeNames[0]);
+                var data = me.harvesterApp.adapter._deserialize(model, chunk.o);
+                var id = chunk.ts.getHighBits() + '_' + chunk.ts.getLowBits();
+                
+                tinySSE.send({id: id, event: routeNames[0] + '_' + chunk.op, data: data})(req, res);
+                next();
+            }
+        });
+
+        docStream.through(consume);
```
Comment (ssebro) [inline lib/multi-sse.js]: how are errors handled here?

File: lib/multi-sse.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    .then(function(query) {
+        var options = {
+            tailable: true,
+            awaitData: true
+        };
+        var stream = me.coll.find(query, options);
+
+        var docStream = hl(function (push, next) {
+            setTimeout(function() {
+                var val = stream.read();
```
Comment (ssebro) [inline lib/multi-sse.js]: Looks like you'll read continuously and rely on highland to allow pausing - is that correct?

File: lib/multi-sse.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                
+                tinySSE.send({id: id, event: routeNames[0] + '_' + chunk.op, data: data})(req, res);
+                next();
+            }
+        });
+
+        docStream.through(consume);
+    })
+    .catch(function(err) {
+        res.end();
```
Comment (ssebro) [inline lib/multi-sse.js]: Definitely log something here

File: lib/multi-sse.js (hunk 4 above)
Comment (mavdi) [inline lib/multi-sse.js]: This line will feed the stream if the value we read isn't null, down the line we won't process the next event unless the one at hand is processed and sent through

**Comment by ssebro** [comment]
:shipit:

**Pull request merged**